### PR TITLE
Migrate multislice PyTorch tests to Airflow

### PR DESCRIPTION
**Pull request by jonb377**
# Description

Migrate multislice PyTorch tests to Airflow.

Also, add support for generated GCS paths in XPK. Unfortunately, task.docker and task.kubernetes execute in an isolated environment without any dependencies available, so we can't directly reference the `metric_config` module to retrieve the value. This requires hard-coding the GCS_OUTPUT environment variable's name into the XPK workload creation command.

# Tests

Ran the tests locally (metrics failures are expected with local airflow): https://screenshot.googleplex.com/BGUoFgA7NBGZ9AA

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run one-shot tests and provided workload links above if applicable. 
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: xlml/utils/xpk.py
Comment (jonb377) [inline xlml/utils/xpk.py:60]: cc @RissyRan @will-cromar in case there's a better approach. `gcs_path` will be an XComArg, so we can't format outside of the `@task.docker`, and `metric_config` is inaccessible from within `@task.docker`.

**Review by RissyRan — COMMENTED** [review]
Thanks Jon!

Just 2 high level questions:

1) what's the error for post_processing for local runs? just curious
2) do we have plans to migrate maxtext tests to use this default folder for each execution in coming PRs? currently, I think tests are using customized gcs folders
3) could you help remove `_` underscore for run_workload?

**Comment by jonb377** [comment]
@RissyRan 

For 1, it's the metrics processing task. There are two errors: one is that it tries to query the Composer environment, and the other is that devboxes don't have write access to the development BigQuery dataset. We could just skip this altogether on local runs, but I hesitate to pull the trigger on that since it would be misleading for a breaking change to metrics processing to pass locally (due to being skipped) but fail in prod.
On 2, I'll work with the test owners on that. Some of the tests have a specific format, since their output is consumed by other teams.
On 3, it's unfortunately due to the `task.docker` and `task.kubernetes` optionality. I think we should migrate this to a raw `@task` to address that, but we need `gcloud` support on the Airflow workers to run XPK. This is a TODO on me right now.

File: xlml/apis/task.py
Comment (jonb377) [reply xlml/apis/task.py:377]: TaskFlow will implicitly add the dependencies, so I left it out of L400. Let me know if I should make it explicit.

For the shared GCS bucket use case, how would we configure the multiple workloads in one XPK task? Would we need a `List[task_test_config]` and `List[gcp_config]` in `XpkTask` definition?

If we wan to keep the two as distinct XpkTask with the same GCS path, we could make the GCS path an optional argument to `run`, and if it's missing then do the generation via `name_format.generate_gcs_folder_location`. So something like:

```python
def run(gcs_path: XComArg[str]):
   ...
   gcs_path = gcs_path or name_format.generate_gcs_folder_location(...)
```

The in cases where we need to link multiple workloads to the same bucket, do something like:

```python
shared_gcs_path = name_format.generate_gcs_folder_location('shared-task-id')
# Task to convert chkpt
get_gke_config(...).run(gcs_path=shared_gcs_path)
# Task to run model
get_gke_config(...).run(gcs_path=shared_gcs_path)
```

**Comment by RissyRan** [comment]
> @RissyRan
> 
> For 1, it's the metrics processing task. There are two errors: one is that it tries to query the Composer environment, and the other is that devboxes don't have write access to the development BigQuery dataset. We could just skip this altogether on local runs, but I hesitate to pull the trigger on that since it would be misleading for a breaking change to metrics processing to pass locally (due to being skipped) but fail in prod. On 2, I'll work with the test owners on that. Some of the tests have a specific format, since their output is consumed by other teams. On 3, it's unfortunately due to the `task.docker` and `task.kubernetes` optionality. I think we should migrate this to a raw `@task` to address that, but we need `gcloud` support on the Airflow workers to run XPK. This is a TODO on me right now.

I see. Thanks! For 1), I will have a try with local testing env this week and see if we could work around.

File: xlml/apis/task.py
Comment (RissyRan) [reply xlml/apis/task.py:377]:
```
TaskFlow will implicitly add the dependencies, so I left it out of L400. Let me know if I should make it explicit.
```

Let's update that as well for less confusion for readers like me (since we already have L400 there). We could specify something like (workload_id, generate_gcs_folder_location) >> run_workload >> wait_for_workload_completion

Right, the proposal makes sense. I am also thinking to try this `xcom_push` and `xcom_pull` in tasks. For instance, inside of this [method](https://github.com/GoogleCloudPlatform/ml-auto-solutions/blob/4a020882beb90c1df8b5595e55a3fb32f429bb8b/xlml/utils/name_format.py#L51),  it queries key first:
* if has value: directly return;
* if no value: generate and return.

If this works, we don't need `shared_gcs_path = name_format.generate_gcs_folder_location('shared-task-id')` and `def run(gcs_path: Optional[XComArg[str]] = None)`.

This is xcom value from `generate_gcs_folder_location` task: https://screenshot.googleplex.com/PoFqg73CUNNHgZr. It's already there.

File: xlml/utils/xpk.py
Comment (jonb377) [reply xlml/utils/xpk.py:60]: I think the solution here is to move this to a BashOperator and make XPK a dependency for the Composer env: https://cloud.google.com/composer/docs/composer-2/write-dags#bash-operator

I'll take this in a separate PR.

**Review by RissyRan — APPROVED** [review]

File: xlml/utils/xpk.py
Comment (jonb377) [reply xlml/utils/xpk.py:60]: Ah that makes sense, I guess xpk is still under rapid development so PyPi can fall behind. I'll stick with `git clone` for now. Will address in this PR: https://github.com/GoogleCloudPlatform/ml-auto-solutions/pull/208

Comment (jonb377) [reply xlml/utils/xpk.py:60]: @RissyRan FYI I'm going to move the GCS support into https://github.com/GoogleCloudPlatform/ml-auto-solutions/pull/208 and rebase this PR once that's landed.

**Comment by RissyRan** [comment]
Back to our previous conversation about gcs_location implementation for multiple tasks. I noticed when querying gcs_location in 2nd test, we need to know benchmark id for 1st test (we cannot make it same as Airflow will throw name conflict error). One example [here](https://screenshot.googleplex.com/3A2gTPswWdsKwtS).

Currently we don't have a use case for GCE to re-use location, so I agree that it's better to move this logic to run() inside of xpk task.

**Comment by jonb377** [comment]
Rebased and ran the tests locally: https://screenshot.googleplex.com/AzdffiP8mKBX7j7

**Pull request merged**
